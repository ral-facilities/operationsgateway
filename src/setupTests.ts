// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import { Action } from 'redux';
import { Record } from './app.types';
import { initialState } from './state/reducers/operationsgateway.reducer';
import { OperationsGatewayState } from './state/state.types';

export let actions: Action[] = [];
export let resetActions = (): void => {
  actions = [];
};
export const getState = (): OperationsGatewayState => initialState;
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

export const testRecords: Record[] = [
  {
    id: '1',
    metadata: {
      dataVersion: '1',
      shotNum: 1,
      timestamp: '1',
      activeArea: '1',
      activeExperiment: '1',
    },
    channels: {
      test1: {
        metadata: {
          dataType: {
            units: 'km',
          },
        },
        data: 1,
      },
    },
  },
  {
    id: '2',
    metadata: {
      dataVersion: '1',
      shotNum: 2,
      timestamp: '2',
      activeArea: '2',
      activeExperiment: '2',
    },
    channels: {
      test2: {
        metadata: {
          dataType: {
            units: 'km',
          },
        },
        data: 2,
      },
    },
  },
  {
    id: '3',
    metadata: {
      dataVersion: '3',
      shotNum: 3,
      timestamp: '3',
      activeArea: '3',
      activeExperiment: '3',
    },
    channels: {
      test3: {
        metadata: {
          dataType: {
            units: 'km',
          },
        },
        data: 3,
      },
    },
  },
];
