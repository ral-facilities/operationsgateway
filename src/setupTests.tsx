// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
// need to mock <canvas> for plotting
import 'jest-canvas-mock';
import {
  Channel,
  FullChannelMetadata,
  ImageChannel,
  Record,
  RecordRow,
  ScalarChannel,
  WaveformChannel,
} from './app.types';
import { Action, PreloadedState, ThunkAction } from '@reduxjs/toolkit';
import { AppStore, RootState, setupStore } from './state/store';
import { initialState as initialConfigState } from './state/slices/configSlice';
import { initialState as initialTableState } from './state/slices/tableSlice';
import { initialState as initialSearchState } from './state/slices/searchSlice';
import { initialState as initialPlotState } from './state/slices/plotSlice';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { QueryClientProvider, QueryClient } from 'react-query';

// this is needed because of https://github.com/facebook/jest/issues/8987
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let mockActualReact;
jest.doMock('react', () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (!mockActualReact) {
    mockActualReact = jest.requireActual('react');
  }
  return mockActualReact;
});

export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getInitialState = (): RootState => ({
  config: initialConfigState,
  table: initialTableState,
  search: initialSearchState,
  plot: initialPlotState,
});
export const dispatch = (
  action: Action | ThunkAction<void, RootState, unknown, Action<string>>
): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getInitialState, null);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

export const hooksWrapperWithProviders = (state = {}) => {
  const testQueryClient = createTestQueryClient();
  const store = setupStore(state);
  const wrapper = ({ children }) => (
    <Provider store={store}>
      <QueryClientProvider client={testQueryClient}>
        {children}
      </QueryClientProvider>
    </Provider>
  );
  return wrapper;
};

export function renderComponentWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const testQueryClient = createTestQueryClient();
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <QueryClientProvider client={testQueryClient}>
          {children}
        </QueryClientProvider>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderComponentWithStore(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => unknown,
      addEventListener: () => unknown,
      removeEventListener: () => unknown,
      addListener: () => unknown,
      removeListener: () => unknown,
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};

export const testChannels: FullChannelMetadata[] = [
  {
    systemName: 'test_1',
    channel_dtype: 'scalar',
    userFriendlyName: 'Test 1',
    significantFigures: 4,
  },
  {
    systemName: 'test_2',
    channel_dtype: 'scalar',
    significantFigures: 2,
    scientificNotation: false,
  },
  {
    systemName: 'test_3',
    channel_dtype: 'scalar',
    significantFigures: 2,
    scientificNotation: true,
  },
];

export const generateRecord = (num: number): Record => {
  const numStr = `${num}`;

  let channel: Channel;

  if (num % 3 === 0) {
    channel = {
      metadata: {
        channel_dtype: 'scalar',
        units: 'km',
      },
      data:
        num < 10
          ? parseFloat(`${num}${num}${num}.${num}`)
          : parseFloat(
              numStr[0] + numStr[1] + numStr[1] + numStr[1] + '.' + numStr[1]
            ),
    } as ScalarChannel;
  } else if (num % 3 === 1) {
    channel = {
      metadata: {
        channel_dtype: 'image',
        horizontalPixels: num,
        horizontalPixelUnits: numStr,
        verticalPixels: num,
        verticalPixelUnits: numStr,
        cameraGain: num,
        exposureTime: num,
      },
      imagePath: numStr,
      thumbnail: numStr,
    } as ImageChannel;
  } else {
    channel = {
      metadata: {
        channel_dtype: 'waveform',
        xUnits: numStr,
        yUnits: numStr,
      },
      waveformId: numStr,
      thumbnail: numStr,
    } as WaveformChannel;
  }

  return {
    id: numStr,
    metadata: {
      dataVersion: numStr,
      shotnum: num,
      timestamp:
        num < 10 ? `2022-01-0${num} 00:00:00` : `2022-01-${num} 00:00:00`,
      activeArea: numStr,
      activeExperiment: numStr,
    },
    channels: {
      [`test_${num}`]: channel,
    },
  };
};

export const testRecords: Record[] = Array.from(Array(3), (_, i) =>
  generateRecord(i + 1)
);

export const generateRecordRow = (num: number) => {
  const record = generateRecord(num);

  const recordRow: RecordRow = {
    timestamp: record.metadata.timestamp,
    shotnum: record.metadata.shotnum,
    activeArea: record.metadata.activeArea,
    activeExperiment: record.metadata.activeExperiment,
  };

  const keys = Object.keys(record.channels);
  keys.forEach((key: string) => {
    const channel: Channel = record.channels[key];
    let channelData;
    const channelDataType = channel.metadata.channel_dtype;

    switch (channelDataType) {
      case 'scalar':
        channelData = (channel as ScalarChannel).data;
        break;
      case 'image':
        channelData = (channel as ImageChannel).thumbnail;
        channelData = (
          <img src={`data:image/jpeg;base64,${channelData}`} alt={key} />
        );
        break;
      case 'waveform':
        channelData = (channel as WaveformChannel).thumbnail;
        channelData = (
          <img src={`data:image/jpeg;base64,${channelData}`} alt={key} />
        );
    }

    recordRow[key] = channelData;
  });

  return recordRow;
};

export const testRecordRows = Array.from(Array(3), (_, i) =>
  generateRecordRow(i + 1)
);
