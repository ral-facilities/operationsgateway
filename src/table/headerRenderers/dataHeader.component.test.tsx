import React from 'react';
import DataHeader, { DataHeaderProps } from './dataHeader.component';
import {
  render,
  RenderResult,
  screen,
  fireEvent,
  act,
} from '@testing-library/react';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import { flushPromises } from '../../setupTests';

describe('Data Header', () => {
  let props: DataHeaderProps;
  const onSort = jest.fn();
  const onClose = jest.fn();
  const handleOnDragEnd = jest.fn();

  const createView = (): RenderResult => {
    return render(
      <table>
        <thead>
          <DragDropContext onDragEnd={handleOnDragEnd}>
            <Droppable droppableId="columns" direction="horizontal">
              {(provided) => (
                <tr {...provided.droppableProps} ref={provided.innerRef}>
                  <DataHeader {...props} />
                </tr>
              )}
            </Droppable>
          </DragDropContext>
        </thead>
      </table>
    );
  };

  beforeEach(() => {
    props = {
      dataKey: 'test',
      sort: {},
      onSort: onSort,
      onClose: onClose,
      label: 'Test',
      icon: function Icon() {
        return <div>Test</div>;
      },
      resizerProps: {},
      index: 0,
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly without sort or filter', () => {
    props.disableSort = true;
    const view = createView();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders correctly with sort but no filter', () => {
    const view = createView();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('calls onClose when close icon is clicked', () => {
    createView();
    const icon = screen.getByLabelText('close test');

    // eslint-disable-next-line testing-library/no-node-access
    fireEvent.click(icon.firstChild);

    expect(onClose).toHaveBeenCalledWith('test');
  });

  it('removes column from display when header is middle clicked', () => {
    createView();
    const header = screen.getByLabelText('test header');
    fireEvent.mouseDown(header, { button: 1 });
    expect(onClose).toHaveBeenCalledWith('test');
  });

  it.todo('renders correctly with filter but no sort');

  it.todo('renders correctly with sort and filter');

  describe('calls the onSort method when label is clicked', () => {
    it('sets asc order', async () => {
      createView();
      await act(async () => {
        screen.getByTestId('sort test').click();
        await flushPromises();
      });

      expect(onSort).toHaveBeenCalledWith('test', 'asc');
    });

    it('sets desc order', async () => {
      props.sort = {
        test: 'asc',
      };

      createView();
      await act(async () => {
        screen.getByTestId('sort test').click();
        await flushPromises();
      });
      expect(onSort).toHaveBeenCalledWith('test', 'desc');
    });

    it('sets null order', async () => {
      props.sort = {
        test: 'desc',
      };

      createView();
      await act(async () => {
        screen.getByTestId('sort test').click();
        await flushPromises();
      });
      expect(onSort).toHaveBeenCalledWith('test', null);
    });
  });

  describe('calls the onSort method when default sort is specified', () => {
    it('sets asc order', () => {
      props.defaultSort = 'asc';

      createView();
      expect(onSort).toHaveBeenCalledWith('test', 'asc');
    });

    it('sets desc order', () => {
      props.defaultSort = 'desc';

      createView();
      expect(onSort).toHaveBeenCalledWith('test', 'desc');
    });
  });
});
