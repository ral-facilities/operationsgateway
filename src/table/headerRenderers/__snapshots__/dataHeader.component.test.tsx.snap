// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Header renders correctly with sort but no filter 1`] = `
<DocumentFragment>
  <table>
    <thead>
      <tr
        data-rbd-droppable-context-id="1"
        data-rbd-droppable-id="columns"
      >
        <th
          aria-describedby="rbd-hidden-text-1-hidden-text-3"
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
          data-rbd-drag-handle-context-id="1"
          data-rbd-drag-handle-draggable-id="test"
          data-rbd-draggable-context-id="1"
          data-rbd-draggable-id="test"
          draggable="false"
          role="columnheader"
          tabindex="0"
        >
          <div
            style="overflow: hidden; flex: 1;"
          >
            <div
              aria-label="test header"
              class="MuiBox-root css-k008qs"
            >
              <div
                class="MuiBox-root css-12z0wuy"
              >
                <div>
                  Test
                </div>
              </div>
              <div
                class="MuiBox-root css-0"
              >
                <span
                  class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                  data-testid="sort test"
                  role="button"
                  tabindex="0"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-xw0cpc-MuiTypography-root"
                  >
                    Test
                  </p>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                    data-testid="ArrowDownwardIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            style="display: flex; flex-direction: row; width: 40px; justify-content: space-between;"
          >
            <div
              aria-label="close test"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1pmilg0-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </div>
            <div
              style="margin-left: 10px; padding-left: 4px; cursor: col-resize;"
            >
              <hr
                class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-ck7sc0-MuiDivider-root"
              />
            </div>
          </div>
        </th>
      </tr>
    </thead>
  </table>
</DocumentFragment>
`;

exports[`Data Header renders correctly without sort or filter 1`] = `
<DocumentFragment>
  <table>
    <thead>
      <tr
        data-rbd-droppable-context-id="0"
        data-rbd-droppable-id="columns"
      >
        <th
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="test"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="test"
          draggable="false"
          role="columnheader"
          tabindex="0"
        >
          <div
            style="overflow: hidden; flex: 1;"
          >
            <div
              aria-label="test header"
              class="MuiBox-root css-k008qs"
            >
              <div
                class="MuiBox-root css-12z0wuy"
              >
                <div>
                  Test
                </div>
              </div>
              <div
                class="MuiBox-root css-0"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-xw0cpc-MuiTypography-root"
                >
                  Test
                </p>
              </div>
            </div>
          </div>
          <div
            style="display: flex; flex-direction: row; width: 40px; justify-content: space-between;"
          >
            <div
              aria-label="close test"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1pmilg0-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </div>
            <div
              style="margin-left: 10px; padding-left: 4px; cursor: col-resize;"
            >
              <hr
                class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-ck7sc0-MuiDivider-root"
              />
            </div>
          </div>
        </th>
      </tr>
    </thead>
  </table>
</DocumentFragment>
`;
